pyi cleanup

bdhirsh opened the pull request:
These are some followups from the first pyi codegen PR. Still maintaining byte-for-byte compatibility in this one.

- Separated `argument_str()` with a pyi flag into two functions, `argument_str()` and `argument_str_pyi()`
- Added a notes section for pyi at the top of `python.py`
- Added a `Python Interface` section that I moved the free-standing pyi functions to

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49057 pyi codegen refactor - no need to group python signatures by overload name
* #49056 pyi codegen - removing byte-for-byte-compatibility hacks (sorting overloads)
* #49055 pyi codegen - removing byte-for-byte compatibility hacks
* **#49054 pyi cleanup**

Differential Revision: [D25410848](https://our.internmc.facebook.com/intern/diff/D25410848)

bdhirsh committed:
pyi cleanup

[ghstack-poisoned]

bdhirsh commented on .jenkins/pytorch/codegen-test.sh:
```diff
@@ -37,7 +37,6 @@ python -m tools.setup_helpers.generate_code \
 mkdir -p "$OUT"/pyi/torch/_C
 mkdir -p "$OUT"/pyi/torch/nn
 python -m tools.pyi.gen_pyi \
-  --declarations-path "$OUT"/torch/share/ATen/Declarations.yaml \
```
I should have taken this out in the original PR

ljk53 replied:
I should have integrated this script with CI properly...

bdhirsh commented on tools/autograd/gen_python_functions.py:
```diff
@@ -228,8 +228,6 @@ def signature_original(f: NativeFunction) -> str:
         if f.func.is_out_fn():
             opname += '_out'
-        # TODO: remove HACK
-        # I think we want to differentiate inplace functions here.. but we currently don't for the arg parser
```
As Ed pointed out, this is actually the right behavior for the python arg parser.

ljk53 reviewed (approved):
Thanks!

bdhirsh committed:
Update on "pyi cleanup"

These are some followups from the first pyi codegen PR. Still maintaining byte-for-byte compatibility in this one.

- Separated `argument_str()` with a pyi flag into two functions, `argument_str()` and `argument_str_pyi()`
- Added a notes section for pyi at the top of `python.py`
- Added a `Python Interface` section that I moved the free-standing pyi functions to


Differential Revision: [D25410848](https://our.internmc.facebook.com/intern/diff/D25410848)

[ghstack-poisoned]

bdhirsh committed:
Update on "pyi cleanup"

These are some followups from the first pyi codegen PR. Still maintaining byte-for-byte compatibility in this one.

- Separated `argument_str()` with a pyi flag into two functions, `argument_str()` and `argument_str_pyi()`
- Added a notes section for pyi at the top of `python.py`
- Added a `Python Interface` section that I moved the free-standing pyi functions to


Differential Revision: [D25410848](https://our.internmc.facebook.com/intern/diff/D25410848)

[ghstack-poisoned]
